src/Nethermind/Nethermind.Blockchain/Processing/BlockProcessor.BlockProductionTransactionPicker.cs

```diff
@@ -0,0 +1,115 @@
+//  but WITHOUT ANY WARRANTY; without even the implied warranty of
+//  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+//  GNU Lesser General Public License for more details.
+// 
+//  You should have received a copy of the GNU Lesser General Public License
+//  along with the Nethermind. If not, see <http://www.gnu.org/licenses/>.
+// 
+
+using System;
+using System.Collections.Generic;
+using Nethermind.Core;
+using Nethermind.Core.Specs;
+using Nethermind.Evm;
+using Nethermind.Int256;
+using Nethermind.Logging;
+using Nethermind.State;
+
+namespace Nethermind.Blockchain.Processing
+{
+    public partial class BlockProcessor
+    {
+        protected class BlockProductionTransactionPicker
+        {
+            private readonly ISpecProvider _specProvider;
+            
+            public BlockProductionTransactionPicker(ISpecProvider specProvider)
+            {
+                _specProvider = specProvider;
+            }
+            
+            public event EventHandler<TxCheckEventArgs>? CheckTransaction;
+
+            public TxCheckEventArgs CanAddTransaction(Block block, Transaction currentTx, IReadOnlySet<Transaction> transactionsInBlock, IStateProvider stateProvider)
+            {
+                TxCheckEventArgs args = new(transactionsInBlock.Count, currentTx, block, transactionsInBlock);
+                
+                long gasRemaining = block.Header.GasLimit - block.GasUsed;
+                
+                // No more gas available in block for any transactions,
+                // the only case we have to really stop
+                if (GasCostOf.Transaction > gasRemaining)
+                {
+                    return args.Set(TxAction.Stop, "Block full");
+                }
+                
+                if (currentTx.SenderAddress is null)
+                {
+                    return args.Set(TxAction.Skip, "Null sender");
+                }
+
+                if (currentTx.GasLimit > gasRemaining)
+                {
+                    return args.Set(TxAction.Skip, $"Not enough gas in block, gas limit {currentTx.GasLimit} > {gasRemaining}");
+                }
+                
+                if (transactionsInBlock.Contains(currentTx))
+                {
+                    return args.Set(TxAction.Skip, "Transaction already in block");
+                }
+
+                UInt256 expectedNonce = stateProvider.GetNonce(currentTx.SenderAddress);
+                if (expectedNonce != currentTx.Nonce)
+                {
+                    return args.Set(TxAction.Skip, $"Invalid nonce - expected {expectedNonce}");
+                }
+
+                UInt256 balance = stateProvider.GetBalance(currentTx.SenderAddress);
+                if (!HasEnoughFounds(currentTx, balance, args, block))
+                {
+                    return args;
+                }
+
+                CheckTransaction?.Invoke(this, args);
```

Comment:
I wonder if we can do it without events...

Reply:
Either this or some weird inheritence, This is actually simpler IMO. We need in AuRa to be able to reject transactions based on POSDAO rules.